Doc/library/sqlite3.rst

```diff
@@ -155,137 +155,9 @@ Reference
 Module functions and constants
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
```

Comment:
Do you think it would be useful to do the following?
1. add sub-sections for functions and constants
2. add a sentence like ``This section lists :mod:`!sqlite3`'s :ref:`module-functions` and :ref:`module-constants`.``

Reply:
> 1\. add sub-sections for functions and constants

:+1: , I also suggested the same thing above hehe

> 2\. add a sentence like `` This section lists :mod:`!sqlite3`'s :ref:`module-functions` and :ref:`module-constants`. ``

Especially if we do the previous, I'm not sure we really need this—it just restates what is already in the section titles and in the ToC in the sidebar. Did you mean it as an alternative?

Reply:
> +1 , I also suggested the same thing above, heh.

Great minds think alike ;)

> Especially if we do the previous, I'm not sure we really need this—it just restates what is already in the section titles and in the ToC in the sidebar. Did you mean it as an alternative?

The idea was both to set expectations on the content of the section (even though the title already does that), and also to provide a link to skip the list of function and jump to the constants without having to manually scroll through.

I was also thinking if this is something we should start doing/recommending in the style guide too :thinking: The sidebar already provides links, but after reaching  the "Module functions and constants" section, it kind of breaks the flow to switch over to the sidebar to find out which section we are in and see/click on the desired sub-section (some themes might make this easier though).  Normally I think I would just scroll through, but if there was a handy inline link I would probably click it.

Reply:
> Great minds think alike ;)

Funny enough, I actually typed out a reference to the same phrase out above, but decided against it before posting (since my mind isn't anywhere near as great as yours) :laughing: 

> The idea was both to set expectations on the content of the section (even though the title already does that), and also to provide a link to skip the list of function and jump to the constants without having to manually scroll through.

Hmm...this is something I do in the [spec section of PEP 639](https://peps.python.org/pep-0639/#specification) (perhaps with excessive detail), but I'm not sure about in the API reference. Unlike a draft PEP, people aren't generally scrolling through it, they rather search or heading jump to specifically what they are looking for. And while "flow" is important in a tutorial, it is not so much in a reference. Per [Diataxis](https://diataxis.fr/reference/#reference-as-description),

> Reference material should be austere and to the point. One hardly *reads* reference material; one *consults* it.

Still, if there was a nested hierarchy of sections, this could still certainly be valuable by providing a concise and cross-referenced summary at the top level, but given all of the sections are standalone underneath the reference, I'm not sure this one-off case necessarily is worth it—particularly under my suggestion which just splits them into two separate sections, since there really isn't much shared commonality between them.

Like with other organizational things this could be something to think about for the future, but if I'd say we discuss and try it out first to see how it works before formally codifying it, if we do.

Reply:
Fair enough :+1: 

> Unlike a draft PEP, people aren't generally scrolling through it, they rather search or heading jump to specifically what they are looking for.

To clarify, my thinking is that people will either ctrl+f or jump to the first header that sounds close to what they are looking for (e.g. "Module functions and constants", rather than the specific sub-section) and once they got close enough they will (have to) scroll through.  I don't expect them to scroll through the whole page.

> I'm not sure this one-off case necessarily is worth it—particularly under my suggestion which just splits them into two separate sections

Agreed.  If this is split into two separate top-level section there is no need (nor place) for an additional sentence that links to functions and constants.  In addition, having these at the top level will make them easier to find/notice.

Reply:
Also, I think _Module functions and attributes_ are better than _Module functions and constants_.

Reply:
The two options are:
1. Split "Module functions and constants" into "Module functions" and "Module constants", both at the same level where "Module functions and constants" is now (so they should be subsubsections with `^^^^^`)
2. Keep "Module functions and constants", and add  "Module functions" and "Module constants" nested inside (so as paragraphs with `"""""`)

Option number 1 is probably better (flat is better than nested).

> Also, I think Module functions and attributes is better than Module functions and constants.

I generally prefer to use "attributes" for classes/instances, and I would generally refer to this as "module-level variables" or "global variables" (even though "global" is somewhat ambiguous).  Constants it technically inaccurate, since they are only constants in spirit, but they are actually variables.

Reply:
> Option number 1 is probably better (flat is better than nested).

+1

> I generally prefer to use "attributes" for classes/instances, and I would generally refer to this as "module-level variables" or "global variables" (even though "global" is somewhat ambiguous). Constants it technically inaccurate, since they are only constants in spirit, but they are actually variables.

I don't like constants either. Personally, I'd lean towards attributes instead of variables, but I have no strong opinion here; either is fine by me.

Reply:
PTAL :)

Reply:
For what its worth, I prefer "module-level constants" since that describes their intended semantics, even if it is of course technically possible modify them, just like we call "private" attributes as such, and the like. [PEP 8 calls them such as well](https://peps.python.org/pep-0008/#constants). But I'm not a Python expert or anything like you two so take that for what you will :laughing:

Reply:
And I'm not sure if other modules use this terminology? Or maybe they do...

Reply:
> And I'm not sure if other modules use this terminology? Or maybe they do...

I'll check current practice before proceeding with landing this.

Reply:
Current practices are:

- Only `logging` has a heading that start with _Module level ..._; everyone else use _Module ..._
- "Constants" is used for upper-case module level attributes (most often used)
- "Attributes" is used for lower-case module level attributes (seldom used)
- "Variables" is not used for module level attributes

Also, note that a lot of modules describe exceptions in their _Module ..._ subsections.

Since the sqlite3 module has a mix of upper-case and lower-case module level attributes, we could either split them up in _Module attributes_ and _Module constants_, or just use one of them (perhaps _Module constants_).

Reply:
> Since the sqlite3 module has a mix of upper-case and lower-case module level attributes, we could either split them up in _Module attributes_ and _Module constants_, or just use one of them (perhaps _Module constants_).

I used _Module constants_. I'm marking this as resolved.